src/webutil/tableState.js

```diff
@@ -0,0 +1,236 @@
+   */
+  +sortName: string,
+  /**
+    A stateful representation of the current order, represented by "asc" or "desc".
+   */
+  +sortOrder: SortOrderTypes,
+  /**
+    A function that changes the currentPage to represent the indicated index.
+   */
+  +setPageIndex: (number) => void,
+  /**
+    A function that sets the rows per page and updates the pageIndex to an equivalent place.
+   */
+  +setRowsPerPage: (number) => void,
+  /**
+    A function that adds a filter to the map of filters or overwrites an existing filter that matches the key.
+    Every filter stored will be applied together as an AND operation (results will only include elements
+    that pass all filters).
+    Sets pageIndex to 0.
+
+    string: A key that represents the filter in the map.
+    function?: A function that returns false if an element should be excluded from the result. Null to remove the filter at the key.
+   */
+  +createOrUpdateFilterFn: (string, null | ((T) => boolean)) => void,
+  /**
+    A function that sorts the data. Current pageIndex is maintained.
+
+    string: A key representing the sort field. It is is used to toggle sort order and increase efficiency.
+    function?: A function that returns the property of the element that should be used in the sort comparator.
+   */
+  +setSortFn: (string, PluckFn<T>) => void,
+|};
+
+/**
+  A builder function for initializing a new TableState for use in a ReactNode.
+  It encapsulates Sort, Pagination, and Filtering.
+ */
+export function useTableState<T>(
+  // The base array. It should ideally be memoized before being passed in.
+  data: Array<T>,
+  initialOptions?: {
+    // The initial rows per page. Defaults to unpaginated.
+    initialRowsPerPage?: number,
+    // Defaults to unsorted (maintains order of the original data).
+    initialSort?: {
+      // The starting order of the initial sort, represented by "asc" or "desc".
+      sortOrder: SortOrderTypes,
+      // A key representing the sort field. It is is used to toggle sort order and increase efficiency.
+      sortName: string,
+      // A function that returns the property of the element that should be used in the sort comparator.
+      sortFn: PluckFn<T>,
+    },
+  }
+): TableState<T> {
+  const optionDefaults = {
+    initialRowsPerPage: data.length,
+    initialSort: {},
+    ...initialOptions,
+  };
+  const sortDefaults = {
+    sortOrder: SortOrders.ASC,
+    sortName: "",
+    sortFn: null,
+    ...optionDefaults.initialSort,
+  };
+  const [sortName, setSortName] = useState(sortDefaults.sortName);
+  const [sortOrder, setSortOrder] = useState<SortOrderTypes>(
+    sortDefaults.sortOrder
+  );
+  const [sortFn, setSortFn] = useState(() => sortDefaults.sortFn);
+  const [pageIndex, setPageIndex] = useState(0);
+  const [rowsPerPage, setRowsPerPageInternal] = useState(
+    optionDefaults.initialRowsPerPage
+  );
+  type FilterFns = {[string]: null | ((T) => boolean)};
```

Comment:
Usually types are defined outside of function bodies, should move it out for consistency sake and readability.

Reply:
accepted